mobilecoind/strategies/drain-accounts.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2018-2021 The MobileCoin Foundation
+
+"""
+The purpose of this strategy is to drain funds from one account to another.
+This is one way to transfer tokens to a fog account, which were created after bootstrap.
+This is the case for minted tokens (which fog distro cannot transfer).
```

Comment:
This is interesting info I didn't quite know and dont %100 grasp. 

What is bootstrap in this context + is it possible to explain more in the comments about why minted tokens can't be transferred in the normal way?

Reply:
okay, so, there is a program called "bootstrap" that creates a ledger out of thin air and makes an origin block with some accounts funded. the keys are represented in this "keyfiles" stuff.

the program at some point was forked when consensus and fog were separate repos, because CD was run out of fog. at some point when they merged, this program became "mc-util-generate-sample-ledger".

but the colloquial name for this program is still bootstrap I guess. sorry if that is confusing, i can use the other name.

bootstrap is still used in CD to generate an origin block for the network. the origin block gets baked into all the nodes and then they start consensing on top of it.

so in the point of time when fog was just like a demo and there were no enclaves, it was possible to bootstrap funds to fog accounts for the purpose of demos, because we just hard coded the fog public key to a constant.

at the point in time when fog became real, it was no longer practical to do that, because we don't know the fog public key until we deploy fog and see what the enclave spits out. but we have to bootstrap before we deploy anything. so it became impossible to bootstrap funds to fog accounts.

there was a network load testing tool called "slam" which got forked and became called "fog-distribution" or "fog-distro". slam takes the bootstrapped ledger and sends payments in parallel as fast as it can to a set of target addresses, for the purpose of load testing a consensus network. we modified it so it could send to fog accounts, and that is more or less fog-distro.

so now we are at a point where we are testing the new custom token ids. and those are only created right now by minting. (I tried bootstrapping the custom token ids but it broke everything and i never managed to figure out why.)

Minting happens after bootstrap, and so fog distro can't see the minted tokens -- fog distro doesn't sync the ledger it just pays from the bootstrapped ledger, so any funds that arrive after bootstrap, it doesn't see.

So I'm trying to build a new way to fund a few fog accounts in CD. The simplest thing I could think of is to use mobilecoind to fund a handful of accounts, so that i can then run fog test client. I'm iterating on this in an internal branch also.

HTH

Reply:
> slam takes the bootstrapped ledger and sends payments in parallel as fast as it can to a set of target addresses, for the purpose of load testing a consensus network.

should clarify:
slam is very janky, it does not sync the ledger, it just starts with a bootstrapped ledger, and it doesn't really have like a database or wallet or anything. it just takes a bunch of account keys, the bootstrapped ledger, matches the outputs to the keys, and then spawns threads and submits transactions in parallel as fast as it can, and it doesn't wait to see if those transactions land even. it can't see because it doesn't sync the ledger.

so this is why it also can't see the minted outputs

Reply:
Got it. That's good context that I figured could go into the comments.

Something like:
```
This script is meant to provide a method for testing transactions with minted tokens. 

Currently the process of using mc-util-generate-sample-ledger to generate an origin block and then sending transactions via mc-fog-distribution is blind to minted tokens because mc-fog-distribution only sends transactions based on initial balances generated in the sample ledger. This will likely change in the future, but for the moment this script makes token minting & minted token transactions visible to fog by first using mobilecoind to create new fog accounts and transferring balances to them.
```

Not necessary to include, just a thought to clarify the purpose of the script.

Reply:
the thing i worry is that there is considerable churn in the way CD works and if I put a comment like this in the code it will just get stale

but maybe that's fine and they will just be like archaeological comments that explain why the script exists whether or not that's still how CD works